### DISCUSSION TITLE
Managing shared parameters across multiple apps

### DISCUSSION THREAD

**Discussion by benmccallum** [aws/aws-dotnet-extensions-configuration#170] [Q&A]
Hi,

We have a situation where we have many apps, but some shared sets of configuration, say the config for a particular feature that app a and app b both use. Ideally we could share the config across apps, as it's be easier if we had to roll a secret or something.

Originally we started to structure our paths in Param Store such that we'd:
* load /all/* (for every app)
* load /app-x/* (where x is the current app)

For a shared feature config, this means logically we'd put that in say /all/feat-foo, but if there's then an app c which shouldn't have access to that feature's config (because it's highly sensitive and we want to limit its visibility), we now have an issue because it'll be visible to app c which has loaded /all/*.

We explored using policies so that /all/feat-foo was restricted for reads to apps a and b. The problem then is when app c starts up and can't read it feat-foo, it will blow up during startup. 

So now we've moved to organising param paths by feature so that an app calls AddSystemManager once per feature. As we expand features my concern is now that this might have a performance drag on our app given so many config providers.

So I guess I'm wondering:
- should I be worried about a perf drain when I've added say 20 AWS config providers on various paths? 
- are there any suggested patterns around this?

It's almost like Param Store needs this shared config concept supported more natively, one global config, then in each app path it's a virtual reference to the shared one to opt-in that app to receive that config.

Thanks in advance for any wisdom anyone can impart :)

**Comment by ashishdhingra** [comment]
@benmccallum Good morning. Looks like **Hybrid Approach with Explicit Opt-In** should work fine.

**Structure:**
```
/app-a/config/*          # App A specific
/app-b/config/*          # App B specific
/shared/feat-foo/*       # Shared feature
/shared/feat-bar/*       # Another shared feature
```
**In each app's configuration:**
```
// App A loads its own config plus opted-in shared features
config.AddSystemsManager("/app-a/config");
config.AddSystemsManager("/shared/feat-foo");  // Explicit opt-in

// App C only loads its own config
config.AddSystemsManager("/app-c/config");
// No access to feat-foo
```
**IAM Policy example:**
```
{
  "Effect": "Allow",
  "Action": "ssm:GetParameter*",
  "Resource": [
    "arn:aws:ssm:region:account:parameter/app-a/*",
    "arn:aws:ssm:region:account:parameter/shared/feat-foo/*"
  ]
}
```

**Pros:**

- Clear security boundaries
- Explicit opt-in prevents accidental exposure
- Easy to audit which apps use which features

**Cons:**

- Requires updating app code when adding new shared features
- More configuration providers as features grow

**Comment by anthony-keller** [comment]
This is the approach we're using, but as the number of different paths increase, we do occasionally hit the request throttle. What would be great is an ability to pass an array of paths in a single request.
